### PR TITLE
解决官方模型flow.decoder转trt fp16的精度问题

### PR DESCRIPTION
**Pull request by lzcchl**
需要安装tensorrt库或者使用含有tensorrt库的镜像，希望能帮到大家，原本想放cosyvoice目录，但是原始docker里面可能没有tensorrt库，实在不知道该放哪，想想还是放这

### CODE REVIEW

**Comment by zmy1116** [comment]
hi,  
sry can't type chinese...
So the descrpancy is caused by the LN layers, not attention related?

**Comment by lzcchl** [comment]
我这么改之后，运行一些语音都不再有nan了，步骤：
step1: 编译.so，这个必须有c++相关环境，我没研究python怎么去编译plugin，maks.sh的一些参数改成你的；
step2: 使用onnx_layernorm_replace.py替换掉onnx中的ln层，改成你的model_path;
step3: 使用build_trt_fp16.py构建engine，推理时也要记得加载.so，和我构建代码中加的几行一样;

**Comment by yuekaizhang** [comment]
@lzcchl  Did you try to use onnx opset > 18 or higher? I think they support to export Layernorm op directly. Then we may don't need to utilize the LN plugin.

**Comment by lzcchl** [comment]
@yuekaizhang 你好，我没有实验onnx opset > 18，其实从现有的 cosyvoice/bin/export_onnx.py + onnx结构看，py中导出试使用参数onnx opset = 18，Layernorm已经是独立层，但它还是使用了opset17域中的Layernorm，预计使用onnx opset > 18不会有什么变化，因为我认为这个问题并不是onnx的而是trt的

**Comment by zmy1116** [comment]
I can confirm that this custom LN plugin helps because I was able to do torch->onnx->trt fp16  without seeing NAN issue, but without the plugin the audio noise is noticeable..  while with the plugin the audio quality matches the orignal model.

[with_cusln.wav](https://github.com/user-attachments/files/24954020/with_cusln.wav)
[without_cusln.wav](https://github.com/user-attachments/files/24954022/without_cusln.wav)

I suppose this resolves the estimator inference issue.   Thanks for the work.

**Comment by lzcchl** [comment]
最近看到autocast这个东西，想来这里能够用到，结果确实也可以解决flow.decoder的精度问题，不依赖plugin，但从我的测试下来，autocast得到的模型速度比plugin的略慢一些。

虽然替换LN plugin能够得到正确结果，但其实是否是LN引发的计算问题，我在使用完autocast之后现在表示怀疑，我想到替换掉它只是因为之前的项目经验，得到正确结果或许只是美丽的巧合，一切都是trt内部在决定算子的使用。

**Comment by aluminumbox** [comment]
> 最近看到autocast这个东西，想来这里能够用到，结果确实也可以解决flow.decoder的精度问题，不依赖plugin，但从我的测试下来，autocast得到的模型速度比plugin的略慢一些。
> 
> 虽然替换LN plugin能够得到正确结果，但其实是否是LN引发的计算问题，我在使用完autocast之后现在表示怀疑，我想到替换掉它只是因为之前的项目经验，得到正确结果或许只是美丽的巧合，一切都是trt内部在决定算子的使用。

use trt version according to requirements, it won't cause nan but fp16 performance is poorer than pytorch amp, maybe trt can fix this problem in later version

**Comment by double12gzh** [comment]
for tritonllm with image `soar97/triton-cosyvoice:25.06`, I can start tritonserver after updating `tensort` version

```bash
root@xxxxxxx:/workspace/CosyVoice# nvcc --version
nvcc: NVIDIA (R) Cuda compiler driver
Copyright (c) 2005-2025 NVIDIA Corporation
Built on Wed_Apr__9_19:24:57_PDT_2025
Cuda compilation tools, release 12.9, V12.9.41
Build cuda_12.9.r12.9/compiler.35813241_0

root@xxxxxxx:/workspace/CosyVoice#pip install --upgrade tensorrt-cu12==10.13.3.9 --extra-index-url https://download.pytorch.org/whl/cu121

root@xxxxxxx:/workspace/CosyVoice# pip list | grep tensor
tensorrt                 10.10.0.31
tensorrt-cu12            10.13.3.9
tensorrt_cu12_bindings   10.13.3.9
tensorrt_cu12_libs       10.13.3.9
tensorrt_llm             0.20.0
```

Also, the following packages are needed:

```bash
apt-get install python3-tk && pip install inflect x-transformers==2.11.24

```

**Comment by double12gzh** [comment]
> 我这么改之后，运行一些语音都不再有nan了，步骤： step1: 编译.so，这个必须有c++相关环境，我没研究python怎么去编译plugin，maks.sh的一些参数改成你的； step2: 使用onnx_layernorm_replace.py替换掉onnx中的ln层，改成你的model_path; step3: 使用build_trt_fp16.py构建engine，推理时也要记得加载.so，和我构建代码中加的几行一样;

Great. I tried your code, works well without `nan`

**Comment by segmentationFaults** [comment]
试了下autocast， 发现溢出的地方是/transformer_blocks.5/attn/MatMul

**Comment by lzcchl** [comment]
@aluminumbox 你好，我在 https://github.com/FunAudioLLM/CosyVoice/blob/main/requirements.txt 中没看到tensorrt版本，请问哪个trt版本下，结果直接就是对的了，后续我会去尝试一下

**Comment by segmentationFaults** [comment]
v100 tensorrt 10.3.0 试了下autocast 之后没有nan了，但是加速太少，和全部fp32 差不多。插件导出之后还是nan。另外我导出的trt 全部都是nan（全部静音），我看你给的音频似乎还是有清晰的语音的。 @lzcchl 请问你那边的trt版本是多少呢？我试试看是不是trt版本的问题

**Comment by lzcchl** [comment]
@segmentationFaults 我使用trtllm25.06镜像中自带的trt 10.10.0.31

**Comment by zenghuiguan** [comment]
> @aluminumbox 你好，我在 https://github.com/FunAudioLLM/CosyVoice/blob/main/requirements.txt 中没看到tensorrt版本，请问哪个trt版本下，结果直接就是对的了，后续我会去尝试一下

https://github.com/FunAudioLLM/CosyVoice/blob/f08872a82fb18b1632c99f4e1d172da08c33d137/requirements.txt#L33

**Comment by lzcchl** [comment]
@zenghuiguan 感谢，我才知道 tensorrt-cu12 其实就是 tensorrt 库，我刚用 tensorrt-cu12==10.13.3.9; sys_platform == 'linux' 尝试，结果是正常的

**Comment by segmentationFaults** [comment]
我用tensorrt-cu12==10.13.3.9 试了下 还是nan☹️， 不过tensorrt-cu12==10.13.3.9  + 插件倒是ok的。。

**Comment by zenghuiguan** [comment]
> @zenghuiguan 感谢，我才知道 tensorrt-cu12 其实就是 tensorrt 库，我刚用 tensorrt-cu12==10.13.3.9; sys_platform == 'linux' 尝试，结果是正常的

不过升到  tensorrt-cu12==10.13.3.9 后，在 teriton_trtllm 中就没有办法运行量化版本了，会直接报错https://github.com/FunAudioLLM/CosyVoice/issues/1831

**Comment by deyituo** [comment]
试了下这个方案可以work，出来的音频不会出现明显的劣化，实现简单速度优化了。具体改下convert_onnx_to_trt，导出fp16 trt时候同时设置下 mixed_precision=True，修改 ONNX（MatMul@FP16, LN/Softmax→FP32），用 STRONGLY_TYPED 构建。

image


```
def prepare_mixed_precision_onnx(onnx_path, output_path, matmul_dtype='fp16'):
    """修改 ONNX 图：MatMul 输入标记为 fp16/bf16，输出 Cast 回 FP32。
    LN/Softmax 保持 FP32。配合 STRONGLY_TYPED 模式构建 TRT engine。
    Args:
        matmul_dtype: 'fp16' 或 'bf16'，MatMul 的计算精度
    """
    import onnx
    import onnx_graphsurgeon as gs
    from onnx import TensorProto

    graph = gs.import_onnx(onnx.load(onnx_path))
    fp32 = np.dtype('float32')
    half = np.dtype('float16')
    cast_to = TensorProto.FLOAT16 if matmul_dtype == 'fp16' else TensorProto.BFLOAT16

    count = 0
    for node in graph.nodes:
        if node.op == 'MatMul':
            for i in range(2):
                inp = node.inputs[i]
                cast_out = gs.Variable(name=f'{node.name}_a{i}_{matmul_dtype}', dtype=half)
                graph.nodes.append(gs.Node(op='Cast', name=f'c{count}i{i}',
                    inputs=[inp], outputs=[cast_out], attrs={'to': cast_to}))
                node.inputs[i] = cast_out
            orig_out = node.outputs[0]
            fp32_out = gs.Variable(name=f'{node.name}_o_fp32', dtype=fp32)
            cast_out_node = gs.Node(op='Cast', name=f'c{count}o',
                inputs=[orig_out], outputs=[fp32_out], attrs={'to': TensorProto.FLOAT})
            graph.nodes.append(cast_out_node)
            for consumer in list(orig_out.outputs):
                if consumer is not cast_out_node:
                    for j, c_inp in enumerate(consumer.inputs):
                        if c_inp is orig_out:
                            consumer.inputs[j] = fp32_out
            count += 1

    logging.info(f'Mixed precision: patched {count} MatMul nodes ({matmul_dtype.upper()} compute, FP32 I/O)')
    graph.cleanup().toposort()
    onnx.save(gs.export_onnx(graph), output_path)
    logging.info(f'Mixed precision ONNX saved to: {output_path}')


def convert_onnx_to_trt(trt_model, trt_kwargs, onnx_model, fp16, mixed_precision=False):
    """构建 TRT engine。
    - mixed_precision=True: 修改 ONNX（MatMul@FP16, LN/Softmax→FP32），用 STRONGLY_TYPED 构建
    - fp16=True: EXPLICIT_BATCH + FP16 flag
    - 否则: EXPLICIT_BATCH (FP32)
    """
    import tensorrt as trt
    logging.info("Converting onnx to trt...")

    use_strongly_typed = False

    # 混合精度：修改 ONNX 图（MatMul@FP16, LN@FP32），用 STRONGLY_TYPED 构建
    if mixed_precision:
        mixed_onnx = onnx_model.replace('.onnx', '.mixed_fp16.onnx')
        if not os.path.exists(mixed_onnx):
            prepare_mixed_precision_onnx(onnx_model, mixed_onnx, matmul_dtype='fp16')
        onnx_model = mixed_onnx
        use_strongly_typed = True

    # STRONGLY_TYPED: TRT 尊重 ONNX 中的类型标注（MatMul@BF16, LN@FP32）
    # EXPLICIT_BATCH: TRT 自己决定精度（FP16 flag 或默认 FP32）
    if use_strongly_typed:
        network_flags = 1 << int(trt.NetworkDefinitionCreationFlag.STRONGLY_TYPED)
    else:
        network_flags = 1 << int(trt.NetworkDefinitionCreationFlag.EXPLICIT_BATCH)

    logger = trt.Logger(trt.Logger.WARNING)
    builder = trt.Builder(logger)
    network = builder.create_network(network_flags)
    parser = trt.OnnxParser(network, logger)
    config = builder.create_builder_config()
    #config.set_memory_pool_limit(trt.MemoryPoolType.WORKSPACE, 1 << 32)  # 4GB

    if not use_strongly_typed and fp16:
        config.set_flag(trt.BuilderFlag.FP16)

    profile = builder.create_optimization_profile()
    with open(onnx_model, "rb") as f:
        if not parser.parse(f.read()):
            for error in range(parser.num_errors):
                logging.info(parser.get_error(error))
            raise ValueError('failed to parse {}'.format(onnx_model))

    for i in range(len(trt_kwargs['input_names'])):
        profile.set_shape(trt_kwargs['input_names'][i], trt_kwargs['min_shape'][i], trt_kwargs['opt_shape'][i], trt_kwargs['max_shape'][i])

    # 非 STRONGLY_TYPED 模式：手动设置 I/O dtype
    if not use_strongly_typed:
        tensor_dtype = trt.DataType.HALF if fp16 else trt.DataType.FLOAT
        for i in range(network.num_inputs):
            network.get_input(i).dtype = tensor_dtype
        for i in range(network.num_outputs):
            network.get_output(i).dtype = tensor_dtype

    config.add_optimization_profile(profile)
    engine_bytes = builder.build_serialized_network(network, config)
    if engine_bytes is None:
        raise ValueError('failed to build trt engine for {}'.format(onnx_model))
    with open(trt_model, "wb") as f:
        f.write(engine_bytes)
    logging.info("Succesfully convert onnx to trt...")
```
